vscode-extensions.vadimcn.vscode-lldb: 1.10.0 -> 1.11.4

FraGag opened the pull request:
* vscode-extensions.vadimcn.vscode-lldb: Fix update.sh (fixes #204311)
* vscode-extensions.vadimcn.vscode-lldb: 1.10.0 -> 1.11.4

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of ~~all binary files (usually in `./result/bin/`)~~ the extension in VSCode
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

FraGag committed:
vscode-extensions.vadimcn.vscode-lldb: Fix update.sh

- Fix nix-shell invocation (`-p packages` and `path` are mutually
  exclusive)
- Patch hashes in the correct files
- Use nurl instead of nix-prefetch (the latter currently fails due to
  <https://github.com/msteen/nix-prefetch/issues/54>)
- Update GitHub repo name

Fixes #204311

FraGag committed:
vscode-extensions.vadimcn.vscode-lldb: 1.10.0 -> 1.11.4

- Remove custom lldb. The extension uses stock lldb since 1.11.0.
- Build the adapter with clang++.
- Build the adapter without weak linkage.
- Fix tools/prep-package.js failing due to missing devDependencies.
- Update GitHub repo name.
- Make various adjustments after upstream reorg.

FraGag commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review rev HEAD -r .`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>vscode-extensions.vadimcn.vscode-lldb</li>
  </ul>
</details>

FraGag commented:
Resolved conflict with d304789e99dfe172d4c9fdb2a621997e204af406.

FraGag commented:
@r4v3n6101 Since you've added yourself as a maintainer for this package, could I trouble you with reviewing this PR when you have time? Also, I can't test on macOS, so if you could verify that the changes work on macOS, that would be a plus. Thanks in advance!

r4v3n6101 commented on pkgs/applications/editors/vscode/extensions/vadimcn.vscode-lldb/update.sh:
```diff
@@ -34,14 +36,14 @@ echo "$old_version -> $version"
 sed -E 's#\bhash = ".*?"#hash = "'$srcHash'"#' --in-place "$nixFile"
-cargoHash=$(nix-prefetch "{ sha256 }: (import $nixpkgs {}).vscode-extensions.vadimcn.vscode-lldb.adapter.cargoDeps.overrideAttrs (_: { outputHash = sha256; })")
-sed -E 's#\bcargoHash = ".*?"#cargoHash = "'$cargoHash'"#' --in-place "$nixFile"
+cargoHash=$(nurl --expr "(import $nixpkgs {}).vscode-extensions.vadimcn.vscode-lldb.adapter.cargoDeps.vendorStaging")
```
Why do we need `nurl` here? Would we leave it like it was before? Or we can't handle like this now?

FraGag replied:
I wrote the reason in the commit message:

```
- Use nurl instead of nix-prefetch (the latter currently fails due to
  <https://github.com/msteen/nix-prefetch/issues/54>)
```

r4v3n6101 commented on pkgs/applications/editors/vscode/extensions/default.nix:
```diff
@@ -5103,7 +5107,17 @@ let
+        # The adapter is meant to be compiled with clang++,
+        # based on the provided CMake toolchain files.
+        # <https://github.com/vadimcn/codelldb/tree/master/cmake>
+        rustPlatform = makeRustPlatform {
```
I guess it would be nice to define it in `vscode-lldb` itself, to prevent `vscode`'s file bloating

FraGag replied:
On one hand, doing this directly in `vscode-lldb` makes it harder to `override` (I guess this is why so many derivations use `override` in this way in `all-packages.nix`). On the other hand, with the introduction of `pkgs/by-name`, perhaps there's a shift away from that pattern (For example, in https://github.com/NixOS/nixpkgs/pull/387764, a bunch of package functions in `pkgs/by-name` changed their `sfml` argument to `sfml_2`. Previously, this would probably have been done with `override`s in `all-packages.nix` or whatever other `.nix` file includes the package).

I'll change it per your suggestion, I also agree that this bloat in `.../vscode/extensions/default.nix` is not pretty. :)

r4v3n6101 commented on pkgs/applications/editors/vscode/extensions/default.nix:
```diff
@@ -5103,7 +5107,17 @@ let
+          stdenv = libcxxStdenv;
+          inherit cargo rustc;
+        };
+        stdenv = libcxxStdenv;
```
`llvmPackages.libcxxStdenv` is better choice, do you think so? The build process will all go around llvm/clang stuff

r4v3n6101 commented on pkgs/applications/editors/vscode/extensions/vadimcn.vscode-lldb/adapter.nix:
```diff
@@ -17,48 +17,47 @@ let
-  buildInputs = lib.optionals stdenv.hostPlatform.isDarwin [ lldb ];
+  # Environment variables, based on <https://github.com/vadimcn/codelldb/blob/master/cargo_config.unix.toml>
+  # The LLDB_* variables are used in adapter/lldb/build.rs.
+  "CC_${LLVM_TRIPLE}" = "${stdenv.cc}/bin/cc";
```
My guess is that these envs must be passed with llvm's stdenv. Try to change `libcxxStdenv` above and have a play w/o these ones. Let me know about the result you get afterwards.

FraGag replied:
Well, `stdenv` here was already overridden to `libcxxStdenv` in `pkgs/applications/editors/vscode/extensions/default.nix`, so these already refer to Clang. I already tried omitting `CC_${LLVM_TRIPLE}` and `CXX_${LLVM_TRIPLE}` with the two `stdenv` overrides in place and it still doesn't work. The error is:

```
g++: error: unrecognized command-line option '-stdlib=libc++'
```

The build script for codelldb's `lldb` crate is passing this option to the C++ compiler assuming it's Clang. Clang recognizes this option but not GCC.

The issue (which I've just now traced) is that if we don't set `CC_${LLVM_TRIPLE}` and `CXX_${LLVM_TRIPLE}`, then the `cc` crate, which is indirectly used by the build script for codelldb's `lldb` crate, will pick the C++ compiler based on the `HOST_CXX` variable, which is set by `cargo-build-hook` using [this value](https://github.com/NixOS/nixpkgs/blob/3c1ae4698778804c49c00d9d15651e0c30cef294/pkgs/build-support/rust/lib/default.nix#L52). The value is derived from `pkgsBuildHost.stdenv.cc`, and the overrides don't apply to _that_ stdenv. On `x86_64-linux`, environment variable `CXX_X86_64_UNKNOWN_LINUX_GNU` is also set, but it looks like `cc` doesn't use it (it looks for `CXX_x86_64-unknown-linux-gnu`, which is what `CXX_${LLVM_TRIPLE}` expands to on `x86_64-linux`, and also for `CXX_x86_64_unknown_linux_gnu`, which has the "wrong" case for its case-sensitive lookup).

`CC_${LLVM_TRIPLE}` is not strictly necessary, since the package doesn't use the C compiler. We could leave it here for consistency with upstream's `cargo_config.unix.toml`, or we could take it away to reduce clutter. `CXX_${LLVM_TRIPLE}` however has to stay. But I think I made a mistake setting `LLVM_TRIPLE` to `stdenv.hostPlatform.config`, I believe it should be `stdenv.buildPlatform.rust.rustcTarget` instead (`buildPlatform` to support cross-compiling, and `rust.rustcTarget` because `config` is not always the correct name for cargo/rustc).

r4v3n6101 reviewed (changes requested):
Hello! Thanks for your work done. Unfortunately, I have no time to review it before. 
I wrote some suggestions with changes. Will be happy to discuss them.

FraGag committed:
vscode-extensions.vadimcn.vscode-lldb: various adjustments following review

- Use `callPackage` instead of `import` for inner packages
- Change `libcxxStdenv` to `llvmPackages_19.libcxxStdenv`
- Move overrides from `.../vscode/extensions/default.nix` to
  `.../vscode/extensions/vadimcn.vscode-lldb/default.nix`
- adapter: Fix LLVM_TRIPLE

FraGag commented:
> Hello! Thanks for your work done. Unfortunately, I have no time to review it before.

No worries, I understand. :)

> I wrote some suggestions with changes. Will be happy to discuss them.

I'm ready for another review. See my replies to your comments and my changes in the new commit. I rebased the branch on current master, but put the changes in a separate commit to make reviewing easier. I can squash the new commit into the previous one later if desired.

r4v3n6101 reviewed (approved):
Hello there again! 
I've tested it locally on `aarch64-darwin` and it's working just fine.
Changes LGTM, thanks for your commitment!
I mark approved

kinnison commented:
Is there more which needs to be done for this, I would really benefit from seeing this merged (and ideally backported to 24.11 since 25.05 is still a good month away)
